src/fs/file.rs

```diff
@@ -432,10 +434,36 @@ cfg_unix! {
             let file = self.file.clone();
             drop(self);
             Arc::try_unwrap(file)
-                .expect("cannot acquire ownership of the file handle after drop")
+                .expect(ARC_TRY_UNWRAP_EXPECT)
                 .into_raw_fd()
         }
     }
+
+    cfg_io_safety! {
+        use crate::os::unix::io::{AsFd, BorrowedFd, OwnedFd};
+
+        impl AsFd for File {
+            fn as_fd(&self) -> BorrowedFd<'_> {
+                self.file.as_fd()
+            }
+        }
+
+        impl From<OwnedFd> for File {
+            fn from(fd: OwnedFd) -> Self {
+                std::fs::File::from(fd).into()
+            }
+        }
+
+        impl From<File> for OwnedFd {
+            fn from(val: File) -> OwnedFd {
+                let file = val.file.clone();
+                drop(val);
+                Arc::try_unwrap(file)
+                    .expect(ARC_TRY_UNWRAP_EXPECT)
+                    .into()
```

Comment:
Rather than duplicating this code (and the message), I'd suggest having a single internal helper function that returns the internal File, and then calling that from both of these trait impls.